File: Leetcode/LC33SearchInRotatedSortedArray.java
public class LC33SearchInRotatedSortedArray {
    public int search(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;

        
        while (start <= end) {
            int middle = (start + end) / 2;
            
            if (nums[middle] == target) {
                return middle;
            }
            
            if (nums[start] <= nums[middle]) {
                if (nums[start] <= target && nums[middle] >= target) {
                    end = middle - 1;
                } else {
                    start = middle + 1;
                }
            } else {
                if (nums[middle] <= target && nums[end] >= target) {
                    start = middle + 1;
                } else {
                    end = middle - 1;
                }
            }
        }
        
        return -1;
    }

    public static void main(String[] args) {
        LC33SearchInRotatedSortedArray lc33SearchInRotatedSortedArray = new LC33SearchInRotatedSortedArray();
        int[] nums = {3, 1};
        System.out.println(lc33SearchInRotatedSortedArray.search(nums, 1));
    }
}

File: Leetcode/LC33SearchInRotatedSortedArray.ts
// binary search for both the number and pivot point
// so we can find the pivot point by checking the difference between big/small
// but then how do we find the number within?
// let's consider the case
function search(nums: number[], target: number): number {
  let left = 0,
    right = nums.length - 1;

  while (left <= right) {
    const mid = Math.floor((right + left) / 2);

    if (nums[mid] === target) {
      return mid;
    }

    if (nums[left] > nums[mid]) { // right side sorted from mid...right
      if (nums[mid] < target && target <= nums[right]) { // left half contains target
        left = mid + 1;
      } else { // right half contains target
        right = mid - 1;
      }
    } else { // left side sorted from left...mid
      if (nums[left] <= target && target < nums[mid]) { // right half contains target
        right = mid - 1;
      } else {
        left = mid + 1;
      }
    }
  }

  return -1;
}
